[PM-7759] - Range slider field

jaasen-livefront opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->
https://bitwarden.atlassian.net/browse/PM-7759

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->
This PR implements a self-contained range slider component with an accompanying number field. It includes an error state for out of bounds input into the number field. The designs did not include copy for the error message.

## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->
|Default|Focus|Error|
|---|---|---|
|![Screenshot 2024-09-09 at 11 09 35 AM](https://github.com/user-attachments/assets/9e3ba686-dbc2-404e-a239-27ab266c1ac6)|![Screenshot 2024-09-09 at 11 09 26 AM](https://github.com/user-attachments/assets/24185493-8dd8-47df-9ae1-b537f6864379)|![Screenshot 2024-09-09 at 11 09 00 AM](https://github.com/user-attachments/assets/eee5f511-ce0b-4237-8e8b-d0ec6bee7605)|


## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

jaasen-livefront commented:
If anyone can help explain why [seemingly random text failure](https://github.com/bitwarden/clients/actions/runs/10778885385/job/29891213805?pr=10966#step:8:3045) is occurring I'll give you 5 ⭐s !

vleague2 commented on libs/components/src/tw-theme.css:
```diff
@@ -178,19 +178,20 @@
 
 @import "./popover/popover.component.css";
 @import "./search/search.component.css";
+@import "../../tools/slider/src/slider.component.css";
```
I don't think we want to be importing from `tools` in the component lib.

audreyality replied:
I think the appropriate place for the component imports is in each client that uses it.

jaasen-livefront replied:
@vleague2 Ah nice catch! I had included this for testing and forgot to remove.

audreyality replied:
📝 It's now imported as a [styleUrl](https://github.com/bitwarden/clients/blob/cb325e167427c00f4752e8a4eb8219b92172dc16/libs/tools/slider/src/slider.component.ts#L23)

vleague2 commented on libs/tools/slider/package.json:
```diff
@@ -0,0 +1,24 @@
+{
```
Does this need its own package.json??

audreyality replied:
👍🏻 We're splitting this into its own module to make its dependencies to and from other modules clear. That'll make it easier to port once the CL has the official version implemented.

vleague2 replied:
I don't think we'd make use of the package for the CL and since this is temporary until we pull it into the CL and will likely only be used in this one design until then, I am not sure if it's worth the effort of having it as a separate package.

jordan-bite replied:
Ok I've removed any files pertaining to use as a package.

audreyality replied:
This was addressed offline. The tools team is maintaining the package for the reason I stated above. 😄

vleague2 commented on libs/tools/slider/src/slider.component.css:
```diff
@@ -0,0 +1,103 @@
+@tailwind components;
+@tailwind utilities;
+
+:root {
```
Since this is global, I would prefer more specific variable names to avoid the chance of collisions. (I.e. more scoped to the slider)

audreyality replied:
❌ I think this CSS file is actually incorrect. It should be using the one I provided in PM-7759. That one was fully vetted by CL and plays nicely with tailwind.

jordan-bite replied:
@audreyality I used the provided css as a reference but I had to heavily modify it to match the designs and have the component function as expected on for `hover`, `focus`, and properly set the track color when the slider or number inputs are manipulated. I was able to maintain the core css with the exception of `@layer`. I attempted to use the component directly in the web app to see if the issue was specific to storybook but ran into a slough of issues with the css so I abandoned that avenue for now. If you have any insight on this issue I would very much appreciate it!

jaasen-livefront replied:
@audreyality I used the provided css as a reference but I had to heavily modify it to match the designs and have the component function as expected on for hover, focus, and properly set the track color when the slider or number inputs are manipulated. I was able to maintain the core css with the exception of @layer. I attempted to use the component directly in the web app to see if the issue was specific to storybook but ran into a slough of issues with the css so I abandoned that avenue for now. If you have any insight on this issue I would very much appreciate it!

audreyality replied:
That CSS 100% covers all of the expected behaviors of the range slider. So long as we get references to the CL's palette variables, it should work. What problems did you run into?

The CSS in the file extends that with chrome support and our tailwind palette. It does seem that the outline piece has stopped working in FF since we cooked that up, though. I debugged it using the [tailwind playground](https://play.tailwindcss.com/Wfjybk35Tl?file=css). You should be able to plop in the CSS to experiment with it. (In fact, it seems they now restrict the shadow DOM, so it's difficult to see where the problem is 😦.)

audreyality replied:
👍🏻 We chatted offline and fixed the CSS bugs.
♻️ @vleague2 - We need to compile the CSS file with tailwind. Is that something you'd be able to help with? We'd file a ticket w/ tech debt and I'll circle back once we're running QA on the generator.

vleague2 replied:
Okay, I will re-review once the updates are pushed up. I'm not familiar with compiling via tailwind, so do you have any more details you could share about what's needed there?

audreyality replied:
We're using `@tailwind` directives at the top of the file that we need to interpret. That lets us scope the changes to specific layers. It addresses your global namespace pollution feedback.

vleague2 commented on libs/tools/slider/src/slider.component.css:
```diff
@@ -0,0 +1,103 @@
+  initial-value: rgb(var(--color-primary-600));
+}
+
+input[type="range"] {
```
Same css comment here -- I don't know that we are using an input of type range anywhere, but I worry about this being global css. Can we scope this to a classname maybe?

audreyality replied:
📝 This is also [related to my comment above](https://github.com/bitwarden/clients/pull/10966#discussion_r1750796498)

jordan-bite replied:
Agreed and updated.

vleague2 commented on libs/tools/slider/src/slider.stories.ts:
```diff
@@ -0,0 +1,64 @@
+  component: SliderComponent,
+  decorators: [
+    moduleMetadata({
+      imports: [SliderComponent, ReactiveFormsModule],
```
The styling looks weird on the bit input, maybe there are more imports you need here?

jaasen-livefront replied:
@vleague2 Are you able to elaborate? I'd love to get the story as close to what appears in reality.

vleague2 replied:
Yeah! I think you will need to import the `FormFieldModule` here in order to get the `bitInput` component to render. (Basically any components used in the creation of your component would need to be imported, I think.)

vleague2 reviewed (commented):
Couple of thoughts on a quick first pass!

audreyality reviewed (commented):
Overall this looks great! Mostly nits questions. I'm excited to see what comes back from the CL team!

jaasen-livefront commented on libs/tools/slider/src/slider.component.html:
```diff
@@ -0,0 +1,45 @@
+      <input
+        toolsSlider
+        #rangeSlider
+        #sliderDirective="toolsSlider"
```
@audreyality This is to allow the template access to `hasError`.

audreyality commented:
> If anyone can help explain why this seemingly random and unrelated [test failure](https://github.com/bitwarden/clients/actions/runs/10778885385/job/29891213805?pr=10966#step:8:3045) is occurring I'll give you 5 ⭐s !

@ttalty is creating a fix for this.

jordan-bite commented:
> > If anyone can help explain why this seemingly random and unrelated [test failure](https://github.com/bitwarden/clients/actions/runs/10778885385/job/29891213805?pr=10966#step:8:3045) is occurring I'll give you 5 ⭐s !
> 
> @ttalty is creating a fix for this in #10969.

⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐

ttalty commented:
> > > If anyone can help explain why this seemingly random and unrelated [test failure](https://github.com/bitwarden/clients/actions/runs/10778885385/job/29891213805?pr=10966#step:8:3045) is occurring I'll give you 5 ⭐s !
> > 
> > 
> > @ttalty is creating a fix for this in #10969.
> 
> ⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐

This has been fixed

jaasen-livefront commented on libs/tools/slider/src/slider.component.ts:
```diff
@@ -0,0 +1,77 @@
+
+  get roundedSliderValue() {
+    return Math.round(this.sliderValueControl.value || 0);
+  }
```
Indeed! I wasn't considering that step size could in practice be less than 1! The reason for defaulting to 0.1 was to improve UX to provide a "smooth" slider.

audreyality replied:
If you want to make the slider more fine-grained in practice, that's cool by me. Just make it clear that's what the code is doing. If you choose to do that, the step input should be used to constrain the output step values.

audreyality replied:
🌱 We might add this back in later. For now, it's ok that the slider jumps from one step to the next.

jaasen-livefront commented on libs/tools/slider/src/slider.component.ts:
```diff
@@ -0,0 +1,77 @@
+    this.setTrackColor(this.sliderValueControl.value);
+  }
+
+  setTrackColor(value: number) {
+    if (!this.sliderEl) {
+      return;
+    }
+
+    const progress = (value / this.max) * 100;
+    this.sliderEl.nativeElement.style.setProperty("--range-fill-value", `${progress}%`);
+  }
```
Totally! We'll need to keep the logic in the component as well as it is also linked to the number field but this is a great suggestion.

audreyality replied:
The number field doesn't have a progress indicator. What am I missing?

jordan-bite replied:
Correct but the progress slider is updated when the number field changes.

audreyality replied:
Hmm... shouldn't the value sync back to the range, which can then perform the calculation internally?

jordan-bite replied:
@audreyality I'll see if I can work it!

jaasen-livefront replied:
@audreyality It's now all in the directive. Much cleaner implementation!

vleague2 commented on libs/tools/slider/src/slider.component.ts:
```diff
@@ -0,0 +1,98 @@
+import { ToolsSliderDirective } from "./slider.directive";
+
+@Component({
+  selector: "bit-slider",
```
I believe the `bit-` prefix is reserved for CL components -- maybe `tools-slider`?

jordan-bite commented on libs/tools/slider/src/slider.component.html:
```diff
@@ -0,0 +1,45 @@
+        value="{{ sliderValueControl.value }}"
+      />
+      <div class="tw-w-full tw-flex tw-justify-between tw-mt-1">
+        <small *ngIf="min !== undefined" role="none" class="tw-text-muted justify-self-start">{{
+          min
+        }}</small>
+        <small *ngIf="max !== undefined" role="none" class="tw-text-muted justify-self-end">{{
+          max
+        }}</small>
```
I've added `[attr.aria-valuexxx]` to the input which should trigger screen readers to announce any changes including the slider value.

jordan-bite commented:
@audreyality @vleague2 I've addressed all the feedback and adjusted the styling to match what was provided in the ticket. I took one quick stab at getting tailwind to process the css without success. I created a [tech debt ticket](https://bitwarden.atlassian.net/browse/PM-11938) to address this which I also included in the story description.

jaasen-livefront commented:
@audreyality @vleague2 (Apologies this was previously posted from my company GH account).  I've addressed all the feedback and adjusted the styling to match what was provided in the ticket. I took one quick stab at getting tailwind to process the css without success. I created a [tech debt ticket](https://bitwarden.atlassian.net/browse/PM-11938) to address this which I also included in the story description.

jaasen-livefront commented:
@audreyality I pushed a commit to the wrong branch, would you min re-approving?

amorask-bitwarden reviewed (dismissed):
Billing file looks good.

audreyality commented:
> @audreyality I pushed a commit to the wrong branch, would you mind re-approving?

Is the billing change ([80aa97b](https://github.com/bitwarden/clients/pull/10966/commits/80aa97bec7f8a5960cf2299d18af1ea9b6f80531)) supposed to be in a different branch? If so, it should be reverted.

vleague2 commented on libs/tools/slider/src/slider.component.css:
```diff
@@ -0,0 +1,114 @@
+
+@layer components {
+  /*********** Baseline, reset styles ***********/
+  input[type="range"] {
```
From what I can understand from the tailwind docs, `@layer` is a way to organize style specificity by layer, and the css defined in each layer is still global. If we update all the `input[type="range"]` selectors to `tools-slider input[type="range"]`, that will scope the css to only range sliders that are children of a `tools-slider` component element.

vleague2 commented on libs/tools/slider/src/slider.component.ts:
```diff
@@ -0,0 +1,72 @@
+  templateUrl: "./slider.component.html",
+  standalone: true,
+  imports: [JslibModule, CommonModule, ReactiveFormsModule, ToolsSliderDirective],
+  styleUrls: ["./slider.component.css"],
```
This is kind of ramble-y thoughts as I am looking at documentation lol

- I think using `styleUrls` may have issues due to our content security policy because Angular injects the styles. Looking at how Oscar implemented the custom css for the search component, he noted in his PR that he did not add the styles in this way due to the CSP. 
- But maybe the `ViewEncapsulation.None` changes the way the styles are applied and maybe they don't get injected in a way that violates the CSP? Docs are unclear on exactly how this works ugh. But it still mentions that styles are injected to the `head` so I feel like we'd run into the same problem....

jaasen-livefront replied:
@vleague2 Indeed `styleUrls` is not ideal. Unfortunately since this component lives in the tools lib we don't have a way to import custom styles like in the component lib which are imported directly into `tw-theme.css`. This is meant more as a stop-gap until the CL team can implement this. I've removed `ViewEncapulation.None` which I think is the opposite of what we're trying to achieve which _is_ encapsulation.

vleague2 reviewed (commented):
Some more thoughts!

jaasen-livefront commented on libs/tools/slider/src/slider.mdx:
```diff
@@ -0,0 +1,23 @@
+
+A custom styled range (aka slider) input `input[type=range]` to allow users to choose a numeric
+value by manipulating a thumb along a track with a defined min and max values. It includes an
+accompying number field for keyboard input. Keyboard input via left and right arrow is also
```
I've updated the wording to be more precise on what I was referring to which is the slider input.

jaasen-livefront commented on libs/tools/slider/src/slider.component.html:
```diff
@@ -0,0 +1,49 @@
+<div class="tw-w-full tw-flex-col tw-gap-2 tw-flex">
+  <div class="tw-flex tw-gap-3">
+    <div class="tw-flex-col tw-pt-4">
```
I think you're correct! I've added an optional label.

jaasen-livefront commented:
Not sure why but the [windows test](https://github.com/bitwarden/clients/actions/runs/10815913925/job/30006508098?pr=10966) keeps failing.

Hinton commented on libs/tools/slider/src/slider.component.css:
```diff
@@ -0,0 +1,114 @@
+@tailwind components;
+@tailwind utilities;
+
+/* Postion of the range's thumb; set by javascript */
+@property --range-fill-value {
+  syntax: "<percentage>";
+  inherits: true;
+  initial-value: 50%;
+}
+
+/* Color that paints the area before the thumb; set by tailwind */
+@property --range-fill-color {
+  syntax: "<color>";
+  inherits: true;
+  initial-value: black;
+}
+
+@property --range-color-bg {
+  syntax: "<color>";
+  inherits: true;
+  initial-value: white;
+}
```
Property does not have broad browser support. Per our browserlist file we support safari 15, will this be correctly transpiled to older CSS?

jaasen-livefront replied:
@audreyality Would you be able to comment on this?

audreyality replied:
It'll be ignored, which is fine. The property is there to ensure the variable is typed, which is important because we're using it to construct other CSS variables.

Hinton commented on libs/tools/slider/src/slider.component.css:
```diff
@@ -0,0 +1,114 @@
+@tailwind base;
+@tailwind components;
+@tailwind utilities;
```
**question**: Is this correct? This would mean we call these multiple times per application. Once for the global tailwind.css and once for the slider. Assuming postcss behaves correctly we will end up with 2x the styles.

audreyality replied:
It's the bare minimum we need for tailwind to compile. I'd much rather have a CL control, but we may never build that. This design is a compromise that gives us tailwind compatibility while limiting the scope to as small a surface area as we can.

audreyality commented:
❌ We're withdrawing this PR. Range support is complex to get right, and we need to prioritize generator UI modernization as a whole. This control is not important enough to further delay the timeline.
